src/WebJobs.Script/Host/HostIdValidator.cs

```diff
@@ -0,0 +1,198 @@
+            LogLevel level = DefaultLevel;
+            string value = _environment.GetEnvironmentVariable(EnvironmentSettingNames.FunctionsHostIdCheckLevel);
+            if (value != null && Enum.TryParse<LogLevel>(value, out LogLevel parsedLevel))
+            {
+                level = parsedLevel;
+            }
+
+            string message = string.Format(Resources.HostIdCollisionFormat, hostId);
+            if (level == LogLevel.Error)
+            {
+                _logger.LogError(message);
+                _applicationLifetime.StopApplication();
+            }
+            else
+            {
+                // we only allow Warning/Error levels to be specified, so anything other than
+                // Error is treated as warning
+                _logger.LogWarning(message);
+            }
+        }
+
+        internal async Task WriteHostIdAsync(string hostId, HostIdInfo hostIdInfo)
+        {
+            try
+            {
+                var containerClient = _storageProvider.GetBlobContainerClient();
+                string blobPath = string.Format(BlobPathFormat, hostId);
+                BlobClient blobClient = containerClient.GetBlobClient(blobPath);
+                BinaryData data = BinaryData.FromObjectAsJson(hostIdInfo);
+                await blobClient.UploadAsync(data);
+            }
+            catch (Exception ex)
+            {
+                // best effort
+                _logger.LogError(ex, "Error writing host ID info");
+            }
+        }
+
+        internal async Task<HostIdInfo> ReadHostIdInfoAsync(string hostId)
+        {
+            HostIdInfo hostIdInfo = null;
+
+            try
+            {
+                // check storage to see if a record already exists for this host ID
+                var containerClient = _storageProvider.GetBlobContainerClient();
+                string blobPath = string.Format(BlobPathFormat, hostId);
+                BlobClient blobClient = containerClient.GetBlobClient(blobPath);
+                var downloadResponse = await blobClient.DownloadAsync();
+                string content;
+                using (StreamReader reader = new StreamReader(downloadResponse.Value.Content))
+                {
+                    content = reader.ReadToEnd();
+                }
+
+                if (!string.IsNullOrEmpty(content))
+                {
+                    hostIdInfo = JsonConvert.DeserializeObject<HostIdInfo>(content);
+                }
+            }
+            catch (RequestFailedException exception) when (exception.Status == 404)
+            {
+                // no record stored for this host ID
+            }
+            catch (Exception ex)
+            {
+                // best effort
+                _logger.LogError(ex, "Error reading host ID info");
+            }
+
+            return hostIdInfo;
+        }
+
+        internal class HostIdInfo
+        {
+            public string Hostname { get; set; }
```

Comment:
Could this also include the untruncated host name so we can present a specific log to the customer about which app has claimed this account?

Reply:
The hostname isn't truncated, only the host ID is.